src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceProgress.cs

```diff
@@ -60,19 +70,60 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
                 _pendingProgress = new PendingProgress();
             }
 
-            _pendingProgress.Update(sourceId, record);
-
             if (_progPane == null)
             {
-                // This is the first time we've received a progress record. Create a progress pane, then show it.
+                // This is the first time we've received a progress record
+                // Create a progress pane
+                // Set up a update flag
+                // Create a timer for updating the flag
 
                 _progPane = new ProgressPane(this);
+
+                if (_progPaneUpdateTimer == null && _progPane != null)
+                {
+                    // Show a progress pane at the first time we've received a progress record
+                    progPaneUpdateFlag = true;
+
+                    // Invoke the timer only once to exclude overlaps
+                    // The timer will be restarted in 'ProgressPaneUpdateTimerElapsed'
+                    _progPaneUpdateTimer = new Timer( new TimerCallback(ProgressPaneUpdateTimerElapsed), null, UpdateTimerThreshold, Timeout.Infinite);
+                }
+            }
+
+            if (progPaneUpdateFlag)
+            {
+                // Update the progress pane only when the timer set up the flag
+                // as a result, we do not block WriteProgress and whole script
+                // and eliminate unnecessary console locks and updates
+                lock (_instanceLock)
```

Comment:
Moving the lock here could cause race conditions. The operations on `_pendingProgress`, `_progPane` and `_progPaneUpdateTimer` outside the lock may interleave with a call to `ResetProgress` on another thread -- another runspace created with the same `PSHost` instance. That would cause unpredictable states to those members.

Is moving this lock really contribute a lot to the performance improvements? It looks to me that lock is to protect the `PSHost` object from being operated on by 2 pipeline threads at the same time, so it shouldn't add any blocking in most cases that Write-Progress is invoked -- creating multiple runspaces with the same `PSHost` object is rare.

Given that, I think we should leave the lock where it was.